### PR TITLE
Leap/user role

### PR DESCRIPTION
**Pull request by simleap**
screenshot https://goo.gl/WnZRmU

### CODE REVIEW

File: app/views/shared/_side_bar.html.haml [hunk 1]
```diff
@@ -22,5 +22,6 @@
             Admin Management
       .card-body
         %ul.nav.flex-column
-          = nav_link 'Users', dashboard_users_path, 'users'
+          - if current_user.is_admin?
+            = nav_link 'Users', dashboard_users_path, 'users'
           = nav_link 'Api-key', dashboard_access_tokens_path, 'access_tokens'
```
Comment (samnang) [inline app/views/shared/_side_bar.html.haml]: Only admin could see API Key.

File: spec/models/user_spec.rb [hunk 2]
```diff
@@ -28,4 +28,18 @@ def assert_validations!
...
     it { assert_validations! }
   end
+
+  describe "defaults" do
+    subject { create(factory) }
+
+    describe "#roles" do
+      it { expect(subject.roles).to eq([:member]) }
+    end
+  end
```
Comment (samnang) [inline spec/models/user_spec.rb]: To me, don't too much fancy with super nested. All you need here:

```rb
it "has `member` role as the default" do
  expect(User.new.riles).to eq([:member])
end 
```

File: spec/models/user_spec.rb [hunk 3]
```diff
@@ -28,4 +28,18 @@ def assert_validations!
...
+
+  it '#is_admin?' do
+    user = create(:user, roles: :admin)
```
Comment (samnang) [inline spec/models/user_spec.rb]: Try to use `User.new` or `build_stubbed` if you justing logic that don't need to touch db at all.

File: spec/system/dashboard/users_spec.rb [hunk 4]
```diff
@@ -1,34 +1,96 @@
...
 
-  it 'show all users from same account' do
-    user2 = create(:user, email: 'bopha@somleng.com.kh', account: user.account)
-    visit '/dashboard/users'
-    expect(page).to have_text(user.email)
-    expect(page).to have_text('bopha@somleng.com.kh')
-  end
+  describe 'only admin can access users pages' do
+    it 'member cannot view user pages' do
+      user.update(roles: :member)
```
Comment (samnang) [inline spec/system/dashboard/users_spec.rb]: It is when I don't like to depend on `before` or `let` too much, it makes your nested spec depend and do something that is not relevant to the current spec too much.

```
context "when a user is not an admin tries to users page" do
  it "redirects them to the default page" do
    user_signed_in

    visit dashboard_users_path
  
	expect(current_page).to eq(dashbaord_path)
  end
end
```

File: spec/system/dashboard/users_spec.rb [hunk 5]
```diff
@@ -1,34 +1,96 @@
...
+
+      visit '/dashboard/users'
```
Comment (samnang) [inline spec/system/dashboard/users_spec.rb]: use rails helper path.

File: spec/system/dashboard/users_spec.rb [hunk 6]
```diff
@@ -1,34 +1,96 @@
...
+
+      expect(page).to have_text('User was successfully updated.')
+      expect(user.roles?(:admin)).to eq true
+    end
+
+    it 'display error message when no roles selected' do
+      visit edit_dashboard_user_path(user)
+
+      uncheck('Member')
+      uncheck('Admin')
```
Comment (samnang) [inline spec/system/dashboard/users_spec.rb]: Should show them as a drop-down instead for now?

File: app/controllers/dashboard/base_controller.rb [hunk 7]
```diff
@@ -4,4 +4,14 @@ class Dashboard::BaseController < ApplicationController
...
   end
+
+  private
+
+  def authorize_admin!
+    deny_access! unless current_user.is_admin?
+  end
+
+  def deny_access!
+    head(:unauthorized)
```
Comment (samnang) [inline app/controllers/dashboard/base_controller.rb]: Should be redirected instead of just empty response coz it's not in API, this one is html request.

**Review by samnang — CHANGES_REQUESTED** [review]
In system spec, try to have your spec to describe scenario as user story. A user story should describe what they want to do and what business value from that. You don't have to have a long text in scenario title.

If you follow that guideline above, your system specs will have less scenarios that don't deliver business values. Make your test suite faster and easy to read and follow as application specification.

File: spec/system/dashboard/users_spec.rb (hunk 6 above)
Comment (simleap) [reply spec/system/dashboard/users_spec.rb]: since we have only two roles and bootstrap doesn't have style for multiple select

**Comment by simleap** [comment]
@samnang, i have refactored this PR and instead of redirect to default page, i have added unauthorized page for unauthorized user.

File: spec/system/dashboard/users_spec.rb [hunk 8]
```diff
@@ -1,34 +1,101 @@
...
+      _user = create(:user, email: 'bopha@somleng.com.kh')
+
+      visit '/dashboard/users'
+
+      expect(page).to have_text(admin.email)
+      expect(page).not_to have_text('bopha@somleng.com.kh')
+    end
+
+    it 'click invite, will open new inviation page' do
       visit 'dashboard/users'
```
Comment (samnang) [inline spec/system/dashboard/users_spec.rb]: Change to rails url helper .

File: spec/system/dashboard/users_spec.rb [hunk 9]
```diff
@@ -1,34 +1,101 @@
...
+      user = create(:user, email: 'bopha@somleng.com.kh', account: admin.account)
+      visit '/dashboard/users'
+      expect(page).to have_text(user.email)
+      expect(page).to have_text('bopha@somleng.com.kh')
+    end
+
+    it 'not show users from other account' do
+      _user = create(:user, email: 'bopha@somleng.com.kh')
+
+      visit '/dashboard/users'
```
Comment (samnang) [inline spec/system/dashboard/users_spec.rb]: Change to rails url helper .

File: spec/system/dashboard/users_spec.rb [hunk 10]
```diff
@@ -1,34 +1,101 @@
...
   end
 
-  context 'click invite user button' do
-    it 'will open new inviation page' do
+  describe 'dashboard list users page' do
+    before { sign_in(admin) }
+
+    it 'show all users from same account' do
+      user = create(:user, email: 'bopha@somleng.com.kh', account: admin.account)
+      visit '/dashboard/users'
```
Comment (samnang) [inline spec/system/dashboard/users_spec.rb]: Change to rails url helper .

**Review by samnang — CHANGES_REQUESTED** [review]
The nav sidebar is break, it shouldn't align in center.

access_denied__401_

Also when I tried to sign in with normal user, it should the unauthorized page. Like I mentioned previously, we should just redirect them to the default page, in this app the default page should be `dashboard_path` and in this context will go to `callouts#index`.

Please ping me if you aren't clear on this.

**Comment by simleap** [comment]
@samnang, it is because current root_path is users#index, redirect user on unauthorized isn't good user experience i think. User will never know what happening.

**Comment by simleap** [comment]
@samnang, i have updated the PR. unauthorized access will redirect user to dashboard root url with an alert flash message.

File: app/controllers/dashboard/access_tokens_controller.rb [hunk 11]
```diff
@@ -1,4 +1,6 @@
 class Dashboard::AccessTokensController < Dashboard::BaseController
+  before_action :authorize_admin!
```
Comment (samnang) [inline app/controllers/dashboard/access_tokens_controller.rb]: Create `Dashboard::AdminController`.

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -28,4 +28,18 @@ def assert_validations!
 
     it { assert_validations! }
   end
+
+  describe "defaults" do
+    subject { create(factory) }
```
Comment (samnang) [inline spec/models/user_spec.rb]: No need, right?

File: spec/system/dashboard/access_tokens_spec.rb [hunk 13]
```diff
@@ -1,29 +1,40 @@
...
+
+  context "when a user is not an admin tries to api key page" do
+    let(:user) { create(:user) }
+
+    it 'redirect to default page with alert message' do
+      sign_in(user)
+
+      visit dashboard_access_tokens_path
+
+      dashboard_root_path
```
Comment (samnang) [inline spec/system/dashboard/access_tokens_spec.rb]: No need this line?

File: spec/system/dashboard/contacts_spec.rb [hunk 14]
```diff
@@ -45,29 +46,39 @@
...
 
     it 'click edit will open edit page' do
-      contact = create(:contact, account: user.account)
+      contact = create(:contact, account: admin.account)
 
+      sign_in admin
       visit dashboard_contact_path(contact)
       click_button 'Edit'
 
       expect(page).to have_current_path(edit_dashboard_contact_path(contact))
```
Comment (samnang) [inline spec/system/dashboard/contacts_spec.rb]: At least should have an assertion with flash message.

**Review by samnang — APPROVED** [review]

**Pull request merged**